src/ConstructionBase.jl

```diff
@@ -24,129 +24,145 @@ for (name, path) in [
+Base.length(@nospecialize(p::PropertyKeys)) = length(values(p))
+_vlength(@nospecialize p) = Val(length(values(p)))
+
+@inline function _getfields(@nospecialize(obj), @nospecialize(fields), v::Val{N}) where {N}
+    ntuple(i -> getfield(obj, getfield(fields, i)), v)
+end
+Base.@propagate_inbounds Base.getindex(@nospecialize(p::PropertyKeys), @nospecialize(i::Integer)) = values(p)[i]
+Base.@propagate_inbounds function Base.getindex(@nospecialize(p1::PropertyIndices), @nospecialize(p2::PropertyIndices))
+    PropertyIndices{_getfields(values(p1), values(p2), _vlength(p2))}()
+end
+function Base.getindex(@nospecialize(p1::PropertyNames), @nospecialize(p2::PropertyIndices))
+    PropertyNames{_getfields(values(p1), values(p2), _vlength(p2))}()
+end
+@inline function Base.iterate(@nospecialize(p::PropertyKeys), state::Int=1)
+    if length(p) > state
+        return nothing
+    else
+        return @inbounds(p[state]), state + 1
     end
 end
 
-getproperties(o::NamedTuple) = o
-getproperties(o::Tuple) = o
-@generated function getproperties(obj)
-    fnames = fieldnames(obj)
-    fvals = map(fnames) do fname
-        Expr(:call, :getproperty, :obj, QuoteNode(fname))
+constructorof(::Type{T}) where {T} = _default_constructorof(T)
+@assume_effects :total function _default_constructorof(T::DataType)
+    getfield(parentmodule(T), getfield(getfield(T, :name), :name))
+end
+
+constructorof(@nospecialize(T::Type{<:Tuple})) = tuple
+constructorof(@nospecialize(T::Type{<:NamedTuple})) = NamedTupleConstructor{fieldnames(T)}()
+
+struct NamedTupleConstructor{names} end
+
+(::NamedTupleConstructor{names})(args...) where {names} = NamedTuple{names}(args)
+
+# getproperties
+getproperties(@nospecialize(obj::NamedTuple)) = obj
+getproperties(@nospecialize(obj::Tuple)) = obj
+getproperties(@nospecialize(obj)) = getproperties(obj, PropertyNames(obj))
+
+function getproperties(@nospecialize(obj::Tuple), @nospecialize(p::PropertyIndices))
+    _getfields(obj, values(p), _vlength(p))
+end
+@inline function getproperties(@nospecialize(obj), @nospecialize(p::PropertyIndices))
+    getproperties(o, PropertyNames(o)[p])
+end
+@generated function getproperties(@nospecialize(obj), ::PropertyNames{nms}) where {nms}
+    t = Expr(:tuple)
+    for n in nms
+        push!(t.args, Expr(:call, :getproperty, :obj, QuoteNode(n)))
     end
-    fvals = Expr(:tuple, fvals...)
-    :(NamedTuple{$fnames}($fvals))
+    Expr(:block, Expr(:meta, :inline), :(NamedTuple{$(nms)}($(t))))
 end
 
 ################################################################################
 ##### setproperties
 ################################################################################
-function setproperties(obj; kw...)
-    setproperties(obj, (;kw...))
+setproperties(obj; kw...) = setproperties(obj, (;kw...))
+setproperties(obj, patch) = isempty(patch) ? obj : _setproperties(obj, patch)
+function _setproperties(@nospecialize(obj::Tuple), @nospecialize(patch::Tuple))
```

Comment:
All the `@nospecialize` are these only for reducing compile time? Or do they also somehow improve inferrability?

Reply:
Mostly for reducing compile time. They shouldn't have an influence on inferrability. We could probably remove them from the untyped arguments for now.

Reply:
I don't have much intuition when `@nospecialized` should be used, your call.